modules/core/src/main/scala/com.snowplowanalytics.snowplow.lakes/processing/Processing.scala

```diff
@@ -151,26 +152,27 @@ object Processing {
   private def parseBytes[F[_]: Async](
     env: Environment[F],
     processor: BadRowProcessor
-  ): Pipe[F, List[Array[Byte]], (List[BadRow], Batched)] =
+  ): Pipe[F, List[ByteBuffer], (List[BadRow], Batched)] =
     _.parEvalMapUnordered(env.cpuParallelism) { list =>
       list
         .traverse { bytes =>
-          Applicative[F].pure {
-            val stringified = new String(bytes, StandardCharsets.UTF_8)
-            Event
-              .parse(stringified)
-              .map(event => Parsed(event, bytes.size, TabledEntity.forEvent(event)))
-              .leftMap { failure =>
-                val payload = BadRowRawPayload(stringified)
-                BadRow.LoaderParsingError(processor, failure, payload)
-              }
-          }
+          for {
+            // order of these byte buffer operations is important
+            numBytes <- Sync[F].delay(bytes.limit() - bytes.position())
+            stringified <- Sync[F].delay(StandardCharsets.UTF_8.decode(bytes).toString)
+          } yield Event
+            .parse(stringified)
+            .map(event => Parsed(event, numBytes, TabledEntity.forEvent(event)))
+            .leftMap { failure =>
+              val payload = BadRowRawPayload(stringified)
+              BadRow.LoaderParsingError(processor, failure, payload)
```

Comment:
Bad rows are never written to a stream and we don't need to check their size right?